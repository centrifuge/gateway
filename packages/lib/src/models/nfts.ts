export class MintNftRequest {
  document_id:string
  proof_fields: string[];
  deposit_address: string;
  registry_address: string
}

export class TransferNftRequest {
  token_id:string;
  to: string;
  registry: string;
}
