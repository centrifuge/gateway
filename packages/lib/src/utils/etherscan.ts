import { BN } from 'ethereumjs-util';

const rootLink = `https://[[network]]etherscan.io`;
// This will be window for browsers and global for node
const scope:any = window || global;
const selectedNetwork = scope['__ETH_NETWORK__'] || 'mainnet';

const subdomain = selectedNetwork === 'mainnet' ? '' : `${selectedNetwork}.`;


export const getTransactionLink = (transaction) => {
  return `${rootLink.replace('[[network]]', subdomain)}/tx/${transaction}`;

};

export const getAddressLink = (address) => {
  return `${rootLink.replace('[[network]]', subdomain)}/address/${address}`;
};

export const getNFTLink = (tokenId, registyAddress) => {
  const tokenToInt = hexToInt(tokenId);
  return `${rootLink.replace('[[network]]', subdomain)}/token/${registyAddress}?a=${tokenToInt}`;
};

export const hexToInt = (hex:string) => {
  return  new BN(hex.replace(/^0x/, ''), 16).toString();
}
