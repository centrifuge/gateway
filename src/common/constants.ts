const API_BASE = '/api';

const baseUsersRoute = `${API_BASE}/users`;

const userRoutes = {
  base: baseUsersRoute,
  login: `${baseUsersRoute}/login`,
  logout: `${baseUsersRoute}/logout`,
  invite: `${baseUsersRoute}/invite`,
  register: `${baseUsersRoute}/register`,
};

export const ROUTES = Object.freeze({
  API_BASE,
  INVOICES: `${API_BASE}/invoices`,
  USERS: userRoutes,
  CONTACTS: `${API_BASE}/contacts`,
  WEBHOOKS: `${API_BASE}/webhooks`,
  PURCHASE_ORDERS: `${API_BASE}/purchase_orders`,
});

export enum PERMISSIONS {
  CAN_FUND_INVOICES = 'can_fund_invoices',
  CAN_CREATE_INVOICES = 'can_create_invoices',
  CAN_MANAGE_USERS = 'can_manage_users',
  CAN_MANAGE_ACCOUNTS = 'can_manage_accounts',
}
