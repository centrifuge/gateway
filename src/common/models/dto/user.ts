export class User {
  constructor(
    readonly username: string,
    readonly password: string,
    readonly id: string,
  ) {}
}
