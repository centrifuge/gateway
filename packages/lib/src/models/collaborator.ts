import { DOCUMENT_ACCESS } from './document';

export class Collaborator {
  address: string;
  name: string;
  access: DOCUMENT_ACCESS;
}

export const collaboratorsToAccessList = (collaborators: Collaborator[], access: string) => {
  return collaborators.filter(
    c => c.access === access,
  ).map(c => c.address);
};
