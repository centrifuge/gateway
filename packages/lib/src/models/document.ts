import {CoreapiCreateDocumentRequest, CoreapiDocumentResponse, CoreapiResponseHeader} from '../centrifuge-node-client';
import { Contact, extendContactLikeObjects } from './contact';
import { Collaborator, collaboratorsToAccessList } from './collaborator';

export interface DocumentRequest extends CoreapiCreateDocumentRequest {
  _id?: string;
}

export interface Document extends CoreapiDocumentResponse {
  ownerId?: string,
  _id?: string,
  fromId?:string,
  createdAt?: Date,
  updatedAt?: Date
  document_id?:string,
  nft_status?:string,
  document_status?:string,
}

export enum DOCUMENT_ACCESS {
  READ = 'read_access',
  WRITE = 'write_access',
}

export const getDocumentCollaborators = (document: Document, contacts: Contact[]) => {
  if (!document || !document.header) return [];
  let userAccess = [];
  Object.values(DOCUMENT_ACCESS).forEach(value => {
    if (document!.header![value]) {
      userAccess = [
        ...userAccess,
        ...(document!.header![value].map(address => {
          return {
            address,
            access: value,
          };
        }) as []),
      ];
    }
  });
  return extendContactLikeObjects(userAccess, contacts);
};

export const createDocumentCollaborators = (collaborators: Collaborator[]) => {
  return Object.values(DOCUMENT_ACCESS).reduce((map, access) => {
    map[access] = collaboratorsToAccessList(collaborators, access);
    return map;
  }, {});
};


