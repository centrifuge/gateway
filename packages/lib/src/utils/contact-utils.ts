import { Contact } from '../models/contact';
import { User } from '../models/user';
import { unionBy } from 'lodash';

export const extendContactsWithUsers = (contacts: Contact[], users: User[]) => {
  return unionBy(contacts, users.map(user => {
    return {
      name: user.name,
      address: user.account,
    };
  }), 'address');
};


export const getContactByAddress = (address: string, contacts: Contact[]) => {
  const contact = contacts.find(c => c.address!.toLowerCase() === address.toLowerCase());
  return contact ? contact : {
    name: address,
    address,
  };
};
