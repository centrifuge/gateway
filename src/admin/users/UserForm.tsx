import React from 'react';
import { Box, Button, FormField, Text, TextInput } from 'grommet';
import { User } from '../../common/models/user';
import * as Yup from 'yup';
import { Formik } from 'formik';
import { PERMISSIONS } from '../../common/constants';
import { MultipleSelect } from '@centrifuge/axis-multiple-select';
import { Schema } from '../../common/models/schema';
import { mapSchemaNames } from '../../common/schema-utils';

type InviteProps = {
  user: User,
  schemas: Schema[],
  onSubmit?: (user) => void;
  onDiscard?: () => void;
}

export default class UserForm extends React.Component<InviteProps> {

  state = { submitted: false };

  onSubmit = (user: User) => {
    this.props.onSubmit && this.props.onSubmit(user);
  };


  render() {

    const userValidation = Yup.object().shape({
      name: Yup.string()
        .max(40, 'Please enter no more than 40 characters')
        .required('This field is required'),
      email: Yup.string()
        .email('Please enter a valid email')
        .required('This field is required'),
      permissions: Yup.array()
        .required('This field is required'),
      schemas: Yup.array()
        .test({
          name: 'test_schemas',
          test: (function(this, value) {
            if (this.parent.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)) {
              return (value && value.length);
            }
            return true;
          }),
          message: 'This field is required',
        }),
    });

    const { user, schemas } = this.props;

    const { submitted } = this.state;

    const permissionOptions = [
      PERMISSIONS.CAN_MANAGE_USERS,
      PERMISSIONS.CAN_MANAGE_SCHEMAS,
      PERMISSIONS.CAN_MANAGE_DOCUMENTS,
      PERMISSIONS.CAN_VIEW_DOCUMENTS,
    ];


    const schemaOptions = schemas.map(i => i.name);

    return (
      <Box margin={{ vertical: 'medium' }}>
        <Formik
          initialValues={user}
          validateOnBlur={submitted}
          validateOnChange={submitted}
          validationSchema={userValidation}
          onSubmit={(values, { setSubmitting }) => {
            if (!values) return;
            this.onSubmit(values);
            setSubmitting(true);
          }}
        >
          {
            ({
               values,
               errors,
               handleChange,
               handleSubmit,
               setFieldValue,
               dirty,
             }) => (
              <form
                onSubmit={event => {
                  this.setState({ submitted: true });
                  handleSubmit(event);
                }}
              >
                <Box gap="medium">
                  <FormField
                    label="Name"
                    error={errors.name}
                  >
                    <TextInput
                      name="name"
                      value={values.name || ''}
                      onChange={handleChange}
                    />
                  </FormField>
                  <FormField
                    label="Email"
                    error={errors!.email}
                  >
                    <TextInput
                      name="email"
                      value={values!.email}
                      onChange={handleChange}
                    />
                  </FormField>

                  <Box margin={{ bottom: 'medium' }}>
                    <FormField
                      label="Permissions"
                      error={errors!.permissions}
                    >
                      <MultipleSelect
                        value={values.permissions}
                        options={permissionOptions}
                        onChange={(selection) => {
                          setFieldValue('permissions', selection);
                          if (!selection.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS)) {
                            setFieldValue('schemas', []);
                          }
                        }}
                      />

                    </FormField>
                  </Box>

                  {
                    values.permissions.includes(PERMISSIONS.CAN_MANAGE_DOCUMENTS) && <>
                      {schemaOptions && schemaOptions.length > 0 ?
                        <Box margin={{ bottom: 'medium' }}>
                          <FormField
                            label="Document schemas"
                            error={errors!.schemas}
                          >
                            <MultipleSelect
                              labelKey={'name'}
                              valueKey={'name'}
                              value={mapSchemaNames(values.schemas, schemas)}
                              options={schemas}
                              onChange={(selection) => {
                                setFieldValue('schemas', selection.map(s => s.name));
                              }}
                            />

                          </FormField>
                        </Box>
                        :
                        <Text color={'status-warning'}>No schemas in the database. Please add schemas!</Text>
                      }
                    </>
                  }

                  <Box direction="row" justify={'end'} gap={'medium'}>
                    <Button
                      label="Discard"
                      onClick={this.props.onDiscard}
                    />
                    <Button
                      disabled={!dirty}
                      type="submit"
                      primary
                      label={user._id ? 'Update' : 'Create'}
                    />
                  </Box>
                </Box>
              </form>
            )
          }
        </Formik>
      </Box>
    );
  }
}

