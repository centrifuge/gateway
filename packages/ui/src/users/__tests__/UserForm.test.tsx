import React from 'react';
import { mount } from 'enzyme';
import UserForm from '../UserForm';
import { User } from '@centrifuge/gateway-lib/models/user';
import { PERMISSIONS } from '@centrifuge/gateway-lib/utils/constants';

it('Should render the create state', () => {
  const user = new User();
  const schemas = [];

  const form = mount(
    <UserForm user={user} schemas={schemas}/>,
  );

  expect(form.html()).toMatchSnapshot();
});


it('Should render in edit state with all fields including values for schemas input', () => {
  const user = new User();
  user.name = 'Test User';
  user.email = 'test@centrifuge.io';
  user.permissions = [PERMISSIONS.CAN_MANAGE_SCHEMAS, PERMISSIONS.CAN_MANAGE_DOCUMENTS];
  user.schemas = ['option-1'];
  const schemas = [{ name: 'option-1' }, { name:  'option-2'}] ;

  const form = mount(
    <UserForm user={user} schemas={schemas}/>,
  );
  expect(form.html()).toMatchSnapshot();
});
