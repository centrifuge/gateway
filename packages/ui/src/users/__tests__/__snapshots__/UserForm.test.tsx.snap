// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render in edit state with all fields including values for schemas input 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 fhDEkk\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Name</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input name=\\"name\\" autocomplete=\\"off\\" class=\\"StyledTextInput-sc-1x30a0s-0 fjRQvG\\" value=\\"Test User\\"></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Email</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input name=\\"email\\" autocomplete=\\"off\\" class=\\"StyledTextInput-sc-1x30a0s-0 fjRQvG\\" value=\\"test@centrifuge.io\\"></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 frudah\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Permissions</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 fcNjcL StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 cGyFlY\\"><div class=\\"StyledBox-sc-13pk1d4-0 gbHXHp\\"><div class=\\"StyledBox-sc-13pk1d4-0 kSBnhK\\"><div class=\\"StyledBox-sc-13pk1d4-0 jUkBCl\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">can_manage_schemas</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gEIaKX\\"></div><div tabindex=\\"0\\" class=\\"StyledBox-sc-13pk1d4-0 cxiFFH\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div><div class=\\"StyledBox-sc-13pk1d4-0 jUkBCl\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">can_manage_documents</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gEIaKX\\"></div><div tabindex=\\"0\\" class=\\"StyledBox-sc-13pk1d4-0 cxiFFH\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 copuvn\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 ojsbV\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 frudah\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Document schemas</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 fcNjcL StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 cGyFlY\\"><div class=\\"StyledBox-sc-13pk1d4-0 gbHXHp\\"><div class=\\"StyledBox-sc-13pk1d4-0 kSBnhK\\"><div class=\\"StyledBox-sc-13pk1d4-0 jUkBCl\\"><span style=\\"line-height: 1;\\" class=\\"StyledText-sc-1sadyjn-0 fWSbXS\\">option-1</span><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 gEIaKX\\"></div><div tabindex=\\"0\\" class=\\"StyledBox-sc-13pk1d4-0 cxiFFH\\"><svg aria-label=\\"Close\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 fjELgu\\"><path fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" d=\\"M3,3 L21,21 M3,21 L21,3\\"></path></svg></div></div></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 copuvn\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 ojsbV\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 epAThu\\"><button type=\\"button\\" class=\\"StyledButton-sc-323bzc-0 cKaKcy\\">Discard</button><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 eRzXCR\\"></div><button disabled=\\"\\" type=\\"submit\\" class=\\"StyledButton-sc-323bzc-0 Fhizx\\">Create</button></div></div></form></div>"`;

exports[`Should render the create state 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 fhDEkk\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Name</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input name=\\"name\\" autocomplete=\\"off\\" class=\\"StyledTextInput-sc-1x30a0s-0 fjRQvG\\" value=\\"\\"></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Email</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input name=\\"email\\" autocomplete=\\"off\\" class=\\"StyledTextInput-sc-1x30a0s-0 fjRQvG\\" value=\\"\\"></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 frudah\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 fRzvEO\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Permissions</label><div class=\\"StyledBox-sc-13pk1d4-0 cHGqW\\"><div class=\\"StyledBox-sc-13pk1d4-0 koruWX\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 fcNjcL StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 cGyFlY\\"><div class=\\"StyledBox-sc-13pk1d4-0 gbHXHp\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" placeholder=\\"Select\\" tabindex=\\"-1\\" type=\\"text\\" class=\\"Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 dQrnKC\\" readonly=\\"\\" value=\\"\\"></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 copuvn\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 ojsbV\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 jwYTgg\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 epAThu\\"><button type=\\"button\\" class=\\"StyledButton-sc-323bzc-0 cKaKcy\\">Discard</button><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 eRzXCR\\"></div><button disabled=\\"\\" type=\\"submit\\" class=\\"StyledButton-sc-323bzc-0 Fhizx\\">Create</button></div></div></form></div>"`;
